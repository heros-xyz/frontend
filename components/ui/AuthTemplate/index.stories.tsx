import type { Meta, StoryObj } from "@storybook/react";
import AuthTemplate from "./index";

const meta: Meta<typeof AuthTemplate> = {
  title: "Components/AuthTemplate",
  component: AuthTemplate,
  tags: ["autodocs"],
  parameters: {
    layout: "fullscreen",
  },
  argTypes: {
    pageType: {
      options: ["athlete", "fan", "signin"],
      control: { type: "select" },
    },
  },
};

export default meta;
type Story = StoryObj<typeof AuthTemplate>;

export const AuthTemplateStory: Story = {
  args: {
    pageType: "athlete",
  },
};
