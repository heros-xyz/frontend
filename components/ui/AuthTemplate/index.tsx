import React, { useEffect, useMemo, useState } from "react";
import {
  Box,
  Button,
  Checkbox,
  Flex,
  Heading,
  Input,
  Text,
} from "@chakra-ui/react";
import { useFormik } from "formik";
import * as Yup from "yup";
import { Else, If, Then } from "react-if";
import Link from "next/link";
import { FaceBookIcon } from "@/components/svg/FaceBook";
import { GoogleIcon } from "@/components/svg/Google";
import ErrorMessage from "@/components/common/ErrorMessage";
import { getAuthErrorCode } from "@/utils/constants";
import { CorporateWebsiteLink, HttpErrorCode } from "@/utils/enums";
import { getWebsiteLink } from "@/utils/link";

interface IAuthProps {
  pageType: string;
  isLoading?: boolean;
  authErrorMessage?: string;
  authErrorCode?: number;
  onSubmitForm: (email: string) => void;
  handleSignInFacebook?: () => void;
  handleSignInGoogle?: () => void;
  defaultEmail?: string;
}

const AuthTemplate: React.FC<IAuthProps> = ({
  pageType = "athlete",
  isLoading,
  authErrorMessage,
  authErrorCode,
  onSubmitForm,
  handleSignInFacebook,
  handleSignInGoogle,
  defaultEmail,
}) => {
  const AuthSchema = Yup.object().shape({
    email: Yup.string()
      .email("Invalid email format.")
      .required("This is a required field.")
      .max(250, "The email address cannot exceed 250 characters.")
      .test("max length before @", "Invalid email format.", (val) => {
        const firstEmail = val?.split("@");
        return Boolean(
          firstEmail?.[0]?.length && firstEmail?.[0]?.length <= 60
        );
      }),
  });

  const formik = useFormik({
    initialValues: {
      email: defaultEmail ?? "",
    },
    validationSchema: AuthSchema,
    onSubmit: (values) => {
      onSubmitForm(values.email);
    },
  });

  const [errorMessage, setErrorMessage] = useState(authErrorMessage);
  const [errorCode, setErrorCode] = useState(authErrorCode);
  const [agreeSignUp, setAgreeSignUp] = useState(false);

  const isAlreadyRegister = useMemo(() => {
    return (
      authErrorMessage === "USER_ALREADY_REGISTERED_OTHER_ROLE" ||
      authErrorMessage === "USER_ALREADY_REGISTERED"
    );
  }, [authErrorMessage]);

  const isNotRegister = useMemo(() => {
    return authErrorMessage === "MUST_SIGN_UP_FIRST";
  }, [authErrorMessage]);

  useEffect(() => {
    setErrorMessage(authErrorMessage);
  }, [authErrorMessage]);

  useEffect(() => {
    setErrorCode(authErrorCode);
  }, [authErrorCode]);

  const renderHeader = () => {
    let title = "";
    let subQuestion = "";
    let signUpLink = "";
    switch (pageType) {
      case "athlete":
        title = "Sign up as an Athlete";
        subQuestion = "Not an athlete?";
        signUpLink = "Sign up as fan here";
        break;
      case "fan":
        title = "Sign up as a Fan";
        subQuestion = "Are you an athlete?";
        signUpLink = "Sign up as an athlete here";
        break;
      case "signin":
        title = "Sign in";
        break;
    }
    return (
      <>
        <Heading
          fontSize={{ base: "4xl", xl: "5xl" }}
          fontWeight="extrabold"
          lineHeight="9"
          mb={{
            base: subQuestion ? "2.5" : "10",
            xl: subQuestion ? "7" : "10",
          }}
        >
          {title}
        </Heading>
        {subQuestion && (
          <Box mb={10}>
            <Text
              as="span"
              mr={2}
              lineHeight="140%"
              fontSize={{ base: "md", xl: "xl" }}
              fontWeight="normal"
            >
              {subQuestion}
            </Text>
            <Link
              href={`/${pageType === "athlete" ? "fan" : "athlete"}/sign-up`}
            >
              <Text
                as="span"
                color="secondary"
                textDecoration="underline"
                fontSize={{ base: "md", xl: "xl" }}
                fontWeight="medium"
                lineHeight="140%"
              >
                {signUpLink}
              </Text>
            </Link>
          </Box>
        )}
      </>
    );
  };

  return (
    <Box
      p={{ base: 5, xl: 0 }}
      bg="primary"
      height="100vh"
      textAlign="center"
      color="white"
      overflowY="auto"
      display="flex"
      justifyContent="center"
    >
      <Box
        mt={{ base: "110px", xl: "100px" }}
        maxWidth={{ base: "100%", lg: "500px" }}
      >
        <Box>
          {renderHeader()}
          <Text
            mb={5}
            fontSize={{ base: "md", xl: "xl" }}
            fontWeight="normal"
            lineHeight={{ base: "100%", xl: "140%" }}
          >
            with socials
          </Text>
          <Flex justifyContent="center" gap={{ base: 2, xl: 8 }} mb={8}>
            <FaceBookIcon
              color="white"
              w={{ base: "33px", xl: "60px" }}
              h={{ base: "33px", xl: "60px" }}
              cursor="pointer"
              onClick={handleSignInFacebook}
            />
            <GoogleIcon
              w={{ base: "33px", xl: "60px" }}
              h={{ base: "33px", xl: "60px" }}
              color="white"
              cursor="pointer"
              onClick={handleSignInGoogle}
            />
          </Flex>
          <Text
            mb={3}
            fontSize={{ base: "md", xl: "xl" }}
            fontWeight="normal"
            lineHeight="140%"
          >
            Or with email
          </Text>
          <form onSubmit={formik.handleSubmit}>
            <Box textAlign="left" mb={8}>
              <Input
                id="email"
                name="email"
                variant="flushed"
                onChange={(e) => {
                  formik.handleChange(e);
                  setErrorMessage("");
                  setErrorCode(undefined);
                }}
                value={formik.values.email}
                placeholder="Email Address"
                fontSize={{ base: "sm", xl: "18px" }}
                fontWeight={{ base: "medium", xl: "normal" }}
                lineHeight="140%"
                _placeholder={{
                  color: "grey.100",
                  fontWeight: 400,
                }}
                autoComplete="off"
              />
              <ErrorMessage
                mt={{ xl: 1 }}
                condition={formik.errors.email && formik.touched.email}
                errorMessage={formik.errors.email}
              />

              {pageType === "signin" && (
                <ErrorMessage
                  mt={{ base: 1.5, xl: 3 }}
                  display="flex"
                  condition={isNotRegister}
                  link="/joining-as"
                  linkText="sign up here!"
                  errorMessage={"You've not registered with this email. Please"}
                />
              )}

              <If condition={pageType !== "signin"}>
                <Then>
                  <Flex mt={6} alignItems="flex-start">
                    <Checkbox
                      size={{ base: "md", xl: "lg" }}
                      onChange={() => setAgreeSignUp(!agreeSignUp)}
                      sx={{
                        span: {
                          borderRadius: 4,
                        },
                      }}
                    />
                    <Box ml={2} fontSize={{ base: "xs", xl: "md" }}>
                      <Text as="span">I agree to </Text>
                      <Link
                        href={getWebsiteLink(
                          CorporateWebsiteLink.TERM_AND_CONDITION
                        )}
                      >
                        <Text
                          as="span"
                          color="secondary"
                          textDecoration="underline"
                        >
                          Herosâ€™s Terms & Conditions
                        </Text>
                      </Link>

                      <Text as="span"> and </Text>
                      <Link
                        href={getWebsiteLink(
                          CorporateWebsiteLink.PRIVACY_POLICY
                        )}
                      >
                        <Text
                          as="span"
                          color="secondary"
                          textDecoration="underline"
                        >
                          Privacy & Cookie Policy
                        </Text>
                      </Link>
                    </Box>
                  </Flex>

                  <ErrorMessage
                    mt={{ base: 1.5, xl: 3 }}
                    display="flex"
                    condition={isAlreadyRegister}
                    link="/sign-in"
                    linkText="sign in here!"
                    errorMessage={
                      "You've already registered with this email. Please"
                    }
                  />
                  <ErrorMessage
                    mt={{ base: 1.5, xl: 3 }}
                    display="flex"
                    condition={isNotRegister}
                    link="/joining-as"
                    linkText="sign up here!"
                    errorMessage={
                      "You've not registered with this email. Please"
                    }
                  />
                </Then>
              </If>
            </Box>

            <Button
              isDisabled={pageType !== "signin" ? !agreeSignUp : false}
              maxWidth={{ base: "full", xl: "250px" }}
              fontSize={{ base: "md", xl: "xl" }}
              isLoading={isLoading}
              _disabled={{
                bg: "grey.100",
                pointerEvents: "none",
              }}
              fontWeight="bold"
              lineHeight="140%"
              bg="secondary"
              color="primary"
              w="100%"
              fontFamily="heading"
              type="submit"
              mb={{ base: "10px", xl: "15px" }}
            >
              Continue with email
            </Button>
          </form>
          <If condition={pageType === "signin"}>
            <Then>
              <Box fontSize={{ base: "xs", xl: "md" }} mb={8}>
                <Text as="span">By signing in you agree to </Text>
                <Link
                  href={getWebsiteLink(CorporateWebsiteLink.TERM_AND_CONDITION)}
                >
                  <Text as="span" color="secondary" textDecoration="underline">
                    Herosâ€™s Terms & Conditions
                  </Text>
                </Link>

                <Text as="span"> and </Text>
                <Link
                  href={getWebsiteLink(CorporateWebsiteLink.PRIVACY_POLICY)}
                >
                  <Text as="span" color="secondary" textDecoration="underline">
                    Privacy & Cookie Policy
                  </Text>
                </Link>
              </Box>
            </Then>
          </If>

          <Box fontSize="md">
            <Text as="span" mr={2} fontWeight="normal" lineHeight="140%">
              {pageType === "signin"
                ? "Don't have an account?"
                : `Already have an account?`}
            </Text>
            <If condition={pageType === "signin"}>
              <Then>
                <Link href="/joining-as">
                  <Text
                    fontFamily="heading"
                    as="span"
                    color="secondary"
                    textDecoration="underline"
                    fontWeight="medium"
                  >
                    Sign up here
                  </Text>
                </Link>
              </Then>
              <Else>
                <Link href="/sign-in">
                  <Text
                    fontFamily="heading"
                    as="span"
                    color="secondary"
                    textDecoration="underline"
                    fontWeight="medium"
                  >
                    Sign in here
                  </Text>
                </Link>
              </Else>
            </If>
          </Box>
        </Box>
      </Box>
    </Box>
  );
};

export default AuthTemplate;
