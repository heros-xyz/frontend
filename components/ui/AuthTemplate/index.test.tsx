import { render } from "@testing-library/react";
import AuthTemplate from ".";

it("renders homepage unchanged", () => {
  const onSubmit = (val: string) => {
    console.log(val);
  };
  const { container } = render(
    <AuthTemplate pageType="athlete" onSubmitForm={onSubmit} />
  );
  expect(container).toMatchSnapshot();
});
