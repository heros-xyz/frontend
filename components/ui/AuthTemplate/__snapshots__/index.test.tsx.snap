// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <div
    class="css-35b87d"
  >
    <div
      class="css-gfl415"
    >
      <div
        class="css-0"
      >
        <h2
          class="chakra-heading css-fgoswi"
        >
          Sign up as an Athlete
        </h2>
        <div
          class="css-1r2f04i"
        >
          <span
            class="chakra-text css-1g108vh"
          >
            Not an athlete?
          </span>
          <a
            href="/fan/sign-up"
          >
            <span
              class="chakra-text css-78p1va"
            >
              Sign up as fan here
            </span>
          </a>
        </div>
        <p
          class="chakra-text css-7fyyeq"
        >
          with socials
        </p>
        <div
          class="css-cfo0ad"
        >
          <svg
            class="chakra-icon css-1mt901b"
            focusable="false"
            viewBox="0 0 33 32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32.5 16C32.5 7.16344 25.3366 0 16.5 0C7.66344 0 0.5 7.16344 0.5 16C0.5 23.9859 6.35094 30.6053 14 31.8056V20.625H9.9375V16H14V12.475C14 8.465 16.3888 6.25 20.0434 6.25C21.7934 6.25 23.625 6.5625 23.625 6.5625V10.5H21.6075C19.62 10.5 19 11.7334 19 13V16H23.4375L22.7281 20.625H19V31.8056C26.6491 30.6053 32.5 23.9859 32.5 16Z"
              fill="currentColor"
            />
          </svg>
          <svg
            class="chakra-icon css-1mt901b"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M16.232 7.32731C15.1486 6.31445 13.7028 5.76268 12.2068 5.78535C9.46935 5.78535 7.14445 7.59562 6.31549 10.0333C5.87596 11.3107 5.87596 12.6939 6.31549 13.9713H6.31934C7.15214 16.4052 9.4732 18.2155 12.2107 18.2155C13.6237 18.2155 14.8369 17.8612 15.7771 17.2354V17.2329C16.8836 16.5148 17.6393 15.3848 17.8745 14.1074H12.2068V10.1467H22.1041C22.2275 10.8345 22.2853 11.5375 22.2853 12.2367C22.2853 15.365 21.1447 18.0099 19.1602 19.8012L19.1623 19.8028C17.4234 21.375 15.0368 22.2858 12.2068 22.2858C8.23941 22.2858 4.61132 20.0938 2.83004 16.6206C1.3418 13.7143 1.34181 10.2903 2.83006 7.38402C4.61133 3.90707 8.23941 1.71505 12.2068 1.71505C14.8132 1.68481 17.3309 2.64476 19.2278 4.39079L16.232 7.32731Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </div>
        <p
          class="chakra-text css-4qsdi6"
        >
          Or with email
        </p>
        <form>
          <div
            class="css-iskddk"
          >
            <input
              class="chakra-input css-1gssj7e"
              id="email"
              name="email"
              placeholder="Email address"
              value=""
            />
            <label
              class="chakra-checkbox css-e2849d"
            >
              <input
                class="chakra-checkbox__input"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-checkbox__control css-dnty2r"
              />
              <span
                class="chakra-checkbox__label css-1oeb2oe"
              >
                <div
                  class="css-e9k318"
                >
                  <span
                    class="chakra-text css-0"
                  >
                    I agree to 
                  </span>
                  <a
                    href="undefined/term-of-service"
                  >
                    <span
                      class="chakra-text css-q4ss7u"
                    >
                      Herosâ€™s Terms & Conditions
                    </span>
                  </a>
                  <span
                    class="chakra-text css-0"
                  >
                     and 
                  </span>
                  <a
                    href="undefined/privacy-policy"
                  >
                    <span
                      class="chakra-text css-q4ss7u"
                    >
                      Privacy & Cookie Policy
                    </span>
                  </a>
                </div>
              </span>
            </label>
          </div>
          <button
            class="chakra-button css-y7y23k"
            disabled=""
            type="submit"
          >
            Continue with email
          </button>
        </form>
        <div
          class="css-504jus"
        >
          <span
            class="chakra-text css-1n8nngx"
          >
            Already have an account?
          </span>
          <a
            href="/sign-in"
          >
            <span
              class="chakra-text css-131fbkz"
            >
              Sign in here
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
